### PR TITLE
Update `WorkspacePath` to support Python 3.12

### PR DESCRIPTION
**Pull request by asnare**
This PR relates to #120 and updates the `WorkspacePath` implementation so that it also works under Python 3.12, in addition to Python 3.10 and Python 3.11.

Changes include:

 - Replacing most of the internal implementation to ensure that the superclass implementations aren't used unless we know they're safe. (They rely on implementations that changed dramatically between 3.11 and 3.12 and are incompatible with each other.)
 - Additional tests to ensure that the public interfaces are tested.

### CODE REVIEW

File: src/databricks/labs/blueprint/paths.py [hunk 1]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+    def __str__(self):
+        try:
+            return self._str
+        except AttributeError:
+            self._str = (self._root + self.parser.sep.join(self._path_parts)) or "."
+            return self._str
+
+    def __bytes__(self):
+        # Super implementations are fine.
+        return super(self).__bytes__()
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:368]: won't it load the super implementation?

File: src/databricks/labs/blueprint/paths.py [hunk 2]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({str(self)!r})"
+
+    def as_uri(
+        self,
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:374]:
```suggestion
        self
```
noise

File: src/databricks/labs/blueprint/paths.py [hunk 3]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
-        """Databricks Workspace works only with absolute paths, so we make sure to
-        return pathlib.Path instead of WorkspacePath to avoid confusion."""
-        return pathlib.PurePath(super().relative_to(*other))
+    @property
+    def parts(self):
+        if self.drive or self.root:
+            parts = (self.drive + self.root, *self._path_parts)
+        else:
+            parts = self._path_parts
+        return parts
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:450]:
```suggestion
            return (self.drive + self.root, *self._path_parts)
        return self._path_parts
```

File: src/databricks/labs/blueprint/paths.py [hunk 4]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+
+    @property
+    def suffix(self):
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:453]: can we get the previous impl, which has more corner cases covered?

**Review by nfx — CHANGES_REQUESTED** [review]
please also look at the downstream failures

File: src/databricks/labs/blueprint/paths.py (hunk 1 above)
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py:368]: There's a different bug here: it should be `super()` instead of `super(self)`.

The superclass will always be loaded, it's really about whether it will invoked and whether this is safe. There are a few categories:
 - _The implementation is in terms of other public APIs_, for all the python versions that we support. These I've marked with the comment you see here. These both act as progress markers to indicate that the implementations have been checked, and when I run the tests with coverage I know I've got a test case that verifies it's fine for all the python versions that we support.
 - _The implementation is in terms of internal `_protected` properties_, which all changed between python 3.11 and 3.12. For these I need an independent implementation, which I'm mainly basing on stripped versions of the default implementation from the (new) abstract base class being introduced in 3.13.
 - _A hybrid of the above._ This I treat like the previous item.

My current strategy is to enumerate the public API (including Dunder methods like this), mark each method, and implement the tests to verify it works across the versions.

File: src/databricks/labs/blueprint/paths.py [hunk 5]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+        if not path_parts:
+            raise ValueError(f"{self!r} has an empty name")
+        path_parts[-1] = name
+        return type(self)(self._ws, self.anchor, *path_parts)
+
+    def with_stem(self, stem):
+        # Super implementations are all fine.
+        return super().with_stem(stem)
+
+    def with_suffix(self, suffix):
```
[the remark this answers is not in the document]
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py:491]: This has always been part of the public API.

File: src/databricks/labs/blueprint/paths.py [hunk 6]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+        stem = self.stem
+        if not stem:
+            msg = f"{self!r} has an empty name"
+            raise ValueError(msg)
+        elif suffix and not suffix.startswith("."):
+            msg = f"{self!r} invalid suffix: {suffix}"
+            raise ValueError(msg)
+        else:
+            return self.with_name(stem + suffix)
```
[the remark this answers is not in the document]
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py:500]: It was; I'm still working my way through the linting problems.

File: src/databricks/labs/blueprint/paths.py [hunk 7]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+        path = self
+        parent = path.parent
+        while path != parent:
+            parents.append(parent)
+            path = parent
+            parent = path.parent
+        return tuple(parents)
+
+    def is_absolute(self):
+        return bool(self.anchor)
```
[the remark this answers is not in the document]
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py:557]: Agreed; that's covered by `test_is_absolute()`

**Comment by asnare** [comment]
> please also look at the downstream failures

Yup, I'm aware of the cause of those; they'll be fixed.

File: src/databricks/labs/blueprint/paths.py [hunk 8]
```diff
@@ -235,66 +277,318 @@ class WorkspacePath(Path):
...
-        drv, root, parts = this._parse_args([path])
-        return this.__from_raw_parts(this, ws, this._flavour, drv, root, parts)
+    def __new__(cls, *args, **kwargs) -> WorkspacePath:
+        # Force all initialisation to go via __init__() irrespective of the (Python-specific) base version.
+        return object.__new__(cls)
+
+    def __init__(self, ws: WorkspaceClient, *args) -> None:  # pylint: disable=super-init-not-called,useless-suppression
+        # We deliberately do _not_ call the super initializer because we're taking over complete responsibility for the
+        # implementation of the public API.
+        raw_paths: list[str] = []
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:287]: can we refactor this into a small staticmethod?

File: src/databricks/labs/blueprint/paths.py [hunk 9]
```diff
@@ -235,66 +277,318 @@ class WorkspacePath(Path):
...
+                path = part
+            case [*parts]:
+                path = cls.parser.join(*parts)
+        if path:
+            root, rel = cls._splitroot(path, sep=cls.parser.sep)
+            # No need to split drv because we don't support it.
+            parsed = tuple(str(x) for x in rel.split(cls.parser.sep) if x and x != ".")
+        else:
+            root, parsed = "", ()
+        return root, parsed
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:337]:
```suggestion
            parsed = tuple(str(x) for x in rel.split(cls.parser.sep) if x and x != ".")
            return root, parsed
        return "", ()
```

File: src/databricks/labs/blueprint/paths.py [hunk 10]
```diff
@@ -235,66 +277,318 @@ class WorkspacePath(Path):
...
+
+    @classmethod
+    def _splitroot(cls, part: str, sep: str) -> tuple[str, str]:
+        # Based on the upstream implementation, with the '//'-specific bit elided because we don't need to
+        # bother with Posix semantics.
+        if part and part[0] == sep:
+            root, path = sep, part.lstrip(sep)
+        else:
+            root, path = "", part
+        return root, path
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:347]:
```suggestion
            return sep, part.lstrip(sep)
        return "", part
```
always return early whenever possible. it's not C :)

File: src/databricks/labs/blueprint/paths.py [hunk 11]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+            msg = f"Invalid name: {name!r}"
+            raise ValueError(msg)
+        path_parts = list(self._path_parts)
+        if not path_parts:
+            raise ValueError(f"{self!r} has an empty name")
+        path_parts[-1] = name
+        return type(self)(self._ws, self.anchor, *path_parts)
+
+    def with_stem(self, stem):
+        # Super implementations are all fine.
```
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/blueprint/paths.py:488]: remove all `useless-parent-delegation` :) even `pylint` yells at this code.

File: src/databricks/labs/blueprint/paths.py [hunk 12]
```diff
@@ -235,66 +277,318 @@ class WorkspacePath(Path):
...
-    def relative_to(self, *other) -> pathlib.PurePath:  # type: ignore
-        """Databricks Workspace works only with absolute paths, so we make sure to
-        return pathlib.Path instead of WorkspacePath to avoid confusion."""
-        return pathlib.PurePath(super().relative_to(*other))
+    @property
+    def _stack(self):
+        return self.anchor, list(reversed(self._path_parts))
+
+    def relative_to(self, other, *more_other, walk_up=False):  # pylint: disable=arguments-differ
+        other = self.with_segments(other, *more_other)
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:507]: how much of this code is copy-pasted from py3.12 impl? we need to properly attribute larger chunks

File: src/databricks/labs/blueprint/paths.py [hunk 13]
```diff
@@ -235,66 +277,318 @@ class WorkspacePath(Path):
...
-        if drv or root:
-            return drv + root + self._flavour.join(parts[1:])
-        return self._flavour.join(parts)
+    def __reduce__(self) -> NoReturn:
+        # Cannot support pickling because we can't pickle the workspace client.
+        msg = "Pickling Workspace paths is not supported."
+        raise NotImplementedError(msg)
+
+    def __fspath__(self):
+        # Cannot support this: Workspace objects aren't accessible via the filesystem.
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:355]: add link to relevant docs

File: src/databricks/labs/blueprint/paths.py [hunk 14]
```diff
@@ -235,66 +277,318 @@ class WorkspacePath(Path):
...
+
+    def __str__(self):
+        try:
+            return self._str
+        except AttributeError:
+            self._str = (self._root + self.parser.sep.join(self._path_parts)) or "."
+            return self._str
+
+    def __bytes__(self):
+        # Super implementations are fine.
```
Comment (nfx) [inline src/databricks/labs/blueprint/paths.py:370]: remove useless override

**Review by nfx — CHANGES_REQUESTED** [review]
let's prioritise the behavior covered in https://github.com/databrickslabs/blueprint/blob/main/tests/integration/test_paths.py and used by UCX. no need to implement 100% pathlib API parity in this release - it can be added in the later increments by community.

File: src/databricks/labs/blueprint/paths.py (hunk 8 above)
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py:300]: Done.

File: src/databricks/labs/blueprint/paths.py (hunk 12 above)
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py:518]: The implementation was heavily inspired by the (default) implementation from the ABC coming in python 3.13. I've since replaced it (and `is_relative_to()`) with an independent implementation.

In general there isn't cut'n'paste code, but the implementations are inspired by the PathLib versions:
 - They're usually simpler because we: a) don't deal with drives; b) assume posix-style paths; c) don't have to deal with case-folding.
 - Because the internal representation is different the implementations are also adapted accordingly. (In particular we store normalised path segments, which eliminates a lot of corner-cases for things like empty or `.` segments.)

File: src/databricks/labs/blueprint/paths.py [hunk 15]
```diff
@@ -235,66 +276,320 @@ class WorkspacePath(Path):
...
+        # language.
+        if not suffix and self.is_notebook():
+            try:
+                suffix = self._SUFFIXES.get(self._object_info.language, "")
+            except DatabricksError:
+                pass
+        return suffix
+
+    @property
+    def suffixes(self):
```
[the remark this answers is not in the document]
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py]: Not particularly. Removed.

File: src/databricks/labs/blueprint/paths.py (hunk 12 above)
Comment (nfx) [reply src/databricks/labs/blueprint/paths.py:497]: ok :) being inspired is sufficient.

File: tests/unit/test_paths.py [hunk 16]
```diff
@@ -13,293 +15,658 @@
...
+
+def test_exists_when_path_exists() -> None:
     ws = create_autospec(WorkspaceClient)
     workspace_path = WorkspacePath(ws, "/test/path")
     ws.workspace.get_status.return_value = True
     assert workspace_path.exists()
 
 
-def test_exists_when_path_does_not_exist():
+def test_exists_when_path_does_not_exist() -> None:
```
[the remark this answers is not in the document]
Comment (asnare) [reply tests/unit/test_paths.py]: Interestingly `mypy` did complain about type issues in a few of the fixtures. I've left the changes to address those complaints in, but removed the `None` return types as suggested to reduce the changes in the PR.

File: tests/unit/test_paths.py [hunk 17]
```diff
@@ -127,22 +524,23 @@ def test_suffix_when_file_has_extension():
...
+    workspace_path._object_info.language = None
     assert workspace_path.suffix == ""
 
 
 def test_suffix_when_file_is_not_notebook():
     ws = create_autospec(WorkspaceClient)
     workspace_path = WorkspacePath(ws, "/test/path")
-    workspace_path.is_notebook = lambda: False
-    assert workspace_path.suffix == ""
+    with patch("databricks.labs.blueprint.paths.WorkspacePath.is_notebook") as mock_is_notebook:
```
Comment (nfx) [inline tests/unit/test_paths.py:541]: https://github.com/databrickslabs/pylint-plugin?tab=readme-ov-file#r8918-explicit-dependency-required

Obscure implicit test dependency with mock.patch(XXX). Rewrite to inject dependencies through constructor.. Using patch to mock dependencies in unit tests can introduce implicit dependencies within a class, making it unclear to other developers. Constructor arguments, on the other hand, explicitly declare dependencies, enhancing code readability and maintainability. However, reliance on patch for testing may lead to issues during refactoring, as updates to underlying implementations would necessitate changes across multiple unrelated unit tests. Moreover, the use of hard-coded strings in patch can obscure which unit tests require modification, as they lack strongly typed references. This coupling of the class under test to concrete classes signifies a code smell, and such code is not easily portable to statically typed languages where monkey patching isn't feasible without significant effort. In essence, extensive patching of external clients suggests a need for refactoring, with experienced engineers recognizing the potential for dependency inversion in such scenarios.

To address this issue, refactor the code to inject dependencies through the constructor. This approach explicitly declares dependencies, enhancing code readability and maintainability. Moreover, it allows for dependency inversion, enabling the use of interfaces to decouple the class under test from concrete classes. This decoupling facilitates unit testing, as it allows for the substitution of mock objects for concrete implementations, ensuring that the class under test behaves as expected. By following this approach, you can create more robust and maintainable unit tests, improving the overall quality of your codebase.

File: tests/unit/test_paths.py [hunk 18]
```diff
@@ -200,18 +598,11 @@ def test_unlink_existing_file():
 def test_unlink_non_existing_file():
     ws = create_autospec(WorkspaceClient)
     workspace_path = WorkspacePath(ws, "/test/path")
-    ws.workspace.get_status.side_effect = NotFound(...)
+    ws.workspace.get_status.side_effect = NotFound("Simulated NotFound")
     with pytest.raises(FileNotFoundError):
         workspace_path.unlink()
 
 
-def test_relative_to():
```
Comment (nfx) [inline tests/unit/test_paths.py:208]: why did you remove existing test?

**Review by nfx — CHANGES_REQUESTED** [review]

File: tests/unit/test_paths.py (hunk 18 above)
Comment (asnare) [reply tests/unit/test_paths.py:208]: There's another version of the same test that covers this (and more situations).

I've moved it here so that this is clearer from the PR diff what the change here really is.

File: tests/unit/test_paths.py (hunk 17 above)
Comment (asnare) [reply tests/unit/test_paths.py]: I've refactored this test so that the behaviour being tested is induced via the (`WorkspaceClient`) mock instead of via monkey-patching.

Assuming this is fine, I've also updated the (related) tests above it to work the same way rather than mutating the state of the object under test.

File: src/databricks/labs/blueprint/paths.py (hunk 13 above)
Comment (asnare) [reply src/databricks/labs/blueprint/paths.py:359]: I've added an explanation and some links. Does this help?

**Review by nfx — APPROVED** [review]
